### PR TITLE
allow empty strings and numbers as metadata type parameters

### PR DESCRIPTION
**Pull request by goetas**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Doc updated   | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/schmittjoh/JMSSerializerBundle/issues/706
| License       | MIT

@Majkl578 do you mind having a look at this?

### CODE REVIEW

File: tests/Serializer/Type/ParserTest.php [hunk 1]
```diff
@@ -72,14 +92,6 @@ public function validTypesProvider(): iterable
             'Foo<"asdf asdf">',
             $type('Foo', ['asdf asdf']),
         ];
-        yield [
```
Comment (goetas) [inline tests/Serializer/Type/ParserTest.php]: removed this as it will not allow `''` (empty strings syntax) and was not available in the 1.x parser

Comment (goetas) [reply tests/Serializer/Type/ParserTest.php]: From what I can guess, what you were trying to do here was handled by YAML

File: src/Type/grammar.pp [hunk 2]
```diff
@@ -2,22 +2,28 @@
 
 %token parenthesis_ <
 %token _parenthesis >
+%token empty_string ""|''
+%token number        (\+|\-)?(0|[1-9]\d*)(\.\d+)?
```
Comment (Majkl578) [inline src/Type/grammar.pp]: This won't allow floats in the following form: `.123`
Should be written as something similar:
```
[+-]?(?:(?:0|[1-9]\d*(?:\.0*[1-9]\d*)?)|\.0*\d*[1-9])
```
https://regex101.com/r/Nadose/1

Also this doesn't handle exponential floats.

File: src/Type/grammar.pp [hunk 3]
```diff
@@ -2,22 +2,28 @@
 
 %token parenthesis_ <
 %token _parenthesis >
+%token empty_string ""|''
```
Comment (Majkl578) [inline src/Type/grammar.pp]: This looks weird, I think this should be handled as part of strings themselves, meaning that they will no longer be a single token but a node instead.

Comment (goetas) [reply src/Type/grammar.pp]: That was the only way I found to make it work with empty strings.
Better solutions are welcome

File: src/Type/grammar.pp (hunk 2 above)
Comment (goetas) [reply src/Type/grammar.pp]: That's true. My goal was to solve the reported bug. A more complete implementation is welcome

File: src/Type/grammar.pp (hunk 3 above)
Comment (Majkl578) [reply src/Type/grammar.pp]: If you make the inner token optional in simple_type (`<quoted_string>?` and `<apostrophed_string>?`), it will work.

Comment (Majkl578) [reply src/Type/grammar.pp]: Of course, you made the token optional (nullable), so the visitor needs to be updated accordingly. :)

**Pull request merged**